lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp

```diff
@@ -222,6 +240,99 @@ static SetVector<Value> getNestedOperands(Operation *op) {
+      if (auto opResult = dyn_cast<OpResult>(operandFromThenBranch)) {
+        auto thenBranchResult = findRootDefiningOp(
+            operandFromThenBranch.getDefiningOp(), opResult.getResultNumber());
+        // add the result to the set.
+        result.insert(thenBranchResult.begin(), thenBranchResult.end());
+      } else if (!dyn_cast<BlockArgument>(operandFromThenBranch)) {
+        result.insert(operandFromThenBranch.getDefiningOp());
+      }
+    }
+    // else branch.
+    {
+      auto operandFromElseBranch = ifOp.thenYield()->getOperand(resultNumber);
+      if (auto opResult = dyn_cast<OpResult>(operandFromElseBranch)) {
+        auto elseBranchResult = findRootDefiningOp(opResult.getDefiningOp(),
+                                                   opResult.getResultNumber());
+        // add the result to the set.
+        result.insert(elseBranchResult.begin(), elseBranchResult.end());
+      } else if (!dyn_cast<BlockArgument>(operandFromElseBranch)) {
+        result.insert(operandFromElseBranch.getDefiningOp());
+      }
+    }
+  } else {
+    result.insert(op);
+  }
+  return std::move(result);
+}
+
+DenseSet<Operation *>
+LoopPipelinerInternal::findRootDefiningOpForSchedulingError(Operation *consumer,
+                                                            Operation *producer,
+                                                            Value operand) {
+  DenseSet<Operation *> rootDefiningOps;
+  if (auto arg = dyn_cast<BlockArgument>(operand)) {
+    auto argNumber = arg.getArgNumber();
+    auto yieldValue = getBlockArgYieldValueFromForLoop(arg);
+    if (!yieldValue) {
+      LDBG("no yield value for arg " << arg << " -> BAIL");
+      return {};
+    }
+
+    if (auto opResult = dyn_cast<OpResult>(*yieldValue)) {
+      auto rootDefiningOp =
+          findRootDefiningOp(producer, opResult.getResultNumber());
+      rootDefiningOps.insert(rootDefiningOp.begin(), rootDefiningOp.end());
+    } else
+      rootDefiningOps.insert(producer);
+  }
+  return std::move(rootDefiningOps);
+}
+
+void LoopPipelinerInternal::printSchedulingError(int64_t distance,
+                                                 Operation *consumer,
+                                                 Operation *producer,
+                                                 Value operand) {
+
+  std::string errorMessage = "operation scheduled before its operands.";
+  std::string likelyBuggyMessage = "This is likely to be a bug. Please "
+                                   "report it.";
+
+  DenseSet<Operation *> rootDefiningOps;
+  if (distance > 0) {
+    rootDefiningOps =
+        findRootDefiningOpForSchedulingError(consumer, producer, operand);
+  }
+  if (rootDefiningOps.empty()) {
+    consumer->emitError() << errorMessage << " " << likelyBuggyMessage;
+  } else {
+    consumer->emitError() << errorMessage;
+    for (auto op : rootDefiningOps) {
+      op->emitError() << "This line likely causes scheduling conflict.";
```

Comment:
Emitting remark should be better here, but I don't know why but remarks won't show up.